src/utilities/applyRequiredStatus.ts

```diff
@@ -0,0 +1,122 @@
+import type { GraphQLOutputType } from '../type/definition';
+import {
+  getNullableType,
+  GraphQLNonNull,
+  isNonNullType,
+  assertListType,
+  GraphQLList,
+  isListType,
+} from '../type/definition';
+import type {
+  SupportArrayNode,
+  NullabilityModifierNode,
+} from '../language/ast';
+import type { ASTReducer } from '../language/visitor';
+import { visit } from '../language/visitor';
+import { GraphQLError } from '../error/GraphQLError';
+
+/**
+ * Implements the "Accounting For Client Controlled Nullability Designators"
+ * section of the spec. In particular, this function figures out the true return
+ * type of a field by taking into account both the nullability listed in the
+ * schema, and the nullability providing by an operation.
+ */
+export function modifiedOutputType(
+  type: GraphQLOutputType,
+  nullabilityNode?: SupportArrayNode | NullabilityModifierNode,
+): GraphQLOutputType {
+  const typeStack: [GraphQLOutputType] = [type];
+
+  const applyStatusReducer: ASTReducer<GraphQLOutputType> = {
+    RequiredDesignator: {
+      leave({ element }) {
+        if (element) {
+          return new GraphQLNonNull(getNullableType(element));
+        }
+
+        // We're working with the inner-most type
+        const nextType = typeStack.pop();
+
+        if (!nextType) {
+          throw new GraphQLError(
+            'List nullability modifier is too deep.',
+            nullabilityNode,
+          );
+        }
+
+        return new GraphQLNonNull(getNullableType(nextType));
+      },
+    },
+    OptionalDesignator: {
+      leave({ element }) {
+        if (element) {
+          return getNullableType(element);
+        }
+
+        // We're working with the inner-most type
+        const nextType = typeStack.pop();
+
+        if (!nextType) {
+          throw new GraphQLError(
+            'List nullability modifier is too deep.',
+            nullabilityNode,
+          );
+        }
+
+        return getNullableType(nextType);
+      },
+    },
+    ListNullabilityDesignator: {
+      enter() {
+        try {
+          const list = assertListType(getNullableType(typeStack.at(-1)));
```

Comment:
Might wanna change this cause it is only available in node `v16` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at and we want to support `v12`+

Reply:
Thanks for pointing that out!